Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.randomcutforest.store;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.CommonUtils.checkState;
+
+import java.util.Arrays;
+
+/**
+ * PointStore is a fixed size repository of points, where each point is a double
+ * array of a specified length. A PointStore counts references to points that
+ * are added, and frees space internally when a given point is no longer in use.
+ * The primary use of this store is to enable compression since the points in
+ * two different trees do not have to be stored separately.
+ *
+ * Stored points are referenced by index values which can be used to look up the
+ * point values and increment and decrement reference counts. Valid index values
+ * are between 0 (inclusive) and capacity (exclusive).
+ */
+public class PointStore {
+
+    private final double[] store;
+    private final int[] refCount;
+    private final int[] freeBlockStack;
+    private int freeBlockPointer;
+    private final int dimensions;
+    private final int capacity;
+
+    /**
+     * Create a new PointStore with the given block size and capacity.
+     *
+     * @param dimensions The number of dimensions in stored points.
+     * @param capacity   The maximum number of points that can be stored.
+     */
+    public PointStore(int dimensions, int capacity) {
+        checkArgument(dimensions > 0, "dimensions must be greater than 0");
+        checkArgument(capacity > 0, "capacity must be greater than 0");
+
+        this.capacity = capacity;
+        this.dimensions = dimensions;
+        store = new double[capacity * dimensions];
+        refCount = new int[capacity];
+        freeBlockStack = new int[capacity];
+
+        for (int j = 0; j < capacity; j++) {
+            freeBlockStack[j] = capacity - j - 1; // reverse order
+        }
+
+        freeBlockPointer = capacity - 1;
+    }
+
+    /**
+     * @return the number of dimensions in stored points for this PointStore.
+     */
+    public int getDimensions() {
+        return dimensions;
+    }
+
+    /**
+     * @return the maximum number of points that can be stored.
+     */
+    public int getCapacity() {
+        return capacity;
+    }
+
+    /**
+     * @return the reference count for the given index.
+     */
+    public int getRefCount(int index) {
+        return refCount[index];
```

Comment:
minor. if the index is not in use, the value can be confusing.

Reply:
I'll update the javadoc for clarity. Currently this method is only used in unit tests, and I think it's convenient to be able to verify that a ref count is 0.